src/mono/wasm/runtime/assets.ts

```diff
@@ -76,72 +126,65 @@ export async function mono_download_assets(): Promise<void> {
     } catch (err: any) {
+        asset.internalPending = undefined;
         if (err && err.status == 404) {
             throw err;
         }
         // second attempt only after all first attempts are queued
         await allDownloadsQueued.promise;
         try {
-            return await start_asset_download_throttle(asset);
+            return await start_asset_download_throttle(asset, downloadData);
         } catch (err) {
+            asset.internalPending = undefined;
             // third attempt after small delay
             await delay(100);
-            return await start_asset_download_throttle(asset);
+            return await start_asset_download_throttle(asset, downloadData);
         }
     }
 }
 
-function resolve_path(asset: AssetEntry, sourcePrefix: string): string {
-    let attemptUrl;
-    const assemblyRootFolder = runtimeHelpers.config.assemblyRootFolder;
-    if (!asset.resolvedUrl) {
-        if (sourcePrefix === "") {
-            if (asset.behavior === "assembly" || asset.behavior === "pdb")
-                attemptUrl = assemblyRootFolder + "/" + asset.name;
-            else if (asset.behavior === "resource") {
-                const path = asset.culture !== "" ? `${asset.culture}/${asset.name}` : asset.name;
-                attemptUrl = assemblyRootFolder + "/" + path;
-            }
-            else {
-                attemptUrl = asset.name;
-            }
-        } else {
-            attemptUrl = sourcePrefix + asset.name;
-        }
-        attemptUrl = runtimeHelpers.locateFile(attemptUrl);
-    }
-    else {
-        attemptUrl = asset.resolvedUrl;
+async function start_asset_download_throttle(asset: AssetEntry, downloadData: boolean): Promise<ArrayBuffer | undefined> {
+    // we don't addRunDependency to allow download in parallel with onRuntimeInitialized event!
+    while (throttling) {
+        await throttling.promise;
     }
-    return attemptUrl;
-}
+    try {
+        ++parallel_count;
+        if (parallel_count == runtimeHelpers.maxParallelDownloads) {
+            if (runtimeHelpers.diagnosticTracing)
+                console.debug("MONO_WASM: Throttling further parallel downloads");
+            throttling = createPromiseController<void>();
+        }
 
-function download_resource(request: ResourceRequest): LoadingResource {
-    if (typeof Module.downloadResource === "function") {
-        const loading = Module.downloadResource(request);
-        if (loading) return loading;
+        const response = await start_asset_download_sources(asset);
+        if (!response) return undefined;
+        if (!downloadData) {
+            return undefined;
+        }
+        return await response.arrayBuffer();
     }
-    const options: any = {};
-    if (request.hash) {
-        options.integrity = request.hash;
+    finally {
+        --parallel_count;
+        if (throttling && parallel_count == runtimeHelpers.maxParallelDownloads - 1) {
```

Comment:
the name throttling sounds like a bool, we should name it something like throttlingFuture